README.md
[![Code Quality](https://landscape.io/github/ssokolow/gtk-recent-scrubber/master/landscape.png)](https://landscape.io/github/ssokolow/gtk-recent-scrubber/master)

GTK Recent Scrubber is a simple little tool for managing your global GTK
recently-used files list.

So far, it will...

1. Watch the GTK recent files list for your desktop session and filter out URIs
   that match a blacklist of prefixes. (Like the
   [HistoryBlock](https://addons.mozilla.org/en-US/firefox/addon/historyblock/)
   extension for Firefox)
2. Make sure your `recently-used.xbel` files are only readable and writable by
   you.
3. Provide a simple `--purge` flag for non-GNOME users who just want to wipe
   everything.

In the future, I also plan to implement a GUI so you can remove individual
entries as well as add support for filtering other lists.

How is this useful? Well, everyone has _some_ guilty pleasure they don't want to
draw attention to. Without this, having a recent files list isn't very useful
because you keep having to clear it.

With it, you can simply blacklist your vices so they don't appear, while the
rest of the list continues to function normally.

Even better, your blacklist is hashed, so it's easier for people to just snoop
around the old fashioned way than to use it as a starting point.

It isn't _technically_ secure, because there is a few-second interval after
programs put things into the list and before they're filtered out again, but it
should be good enough for most people.

## Requirements

- Python 3.x
- GTK 3.x (for `Gtk.RecentManager`)
- PyGObject with the GIR definitions for Gtk, Gdk, and GLib

**Debian and derivatives (Ubuntu, Mint, etc.):**

```sh
sudo apt-get install python3 python3-gi gir1.2-glib-2.0 gir1.2-gtk-3.0
```

**Fedora and derivatives:**

```sh
sudo dnf install python3 python3-gobject gtk3
```

## Installation

1. Put `gtk_cleanup.py` file wherever you want and name it whatever you want.
2. Mark it executable. (i.e. `chmod +x gtk_cleanup.py`)
3. Run `gtk_cleanup.py --add <URI or path>` to build your blacklist.
4. Run `gtk_cleanup.py -vv --once` to test it.
5. Use whatever means you normally would to make `gtk_cleanup.py` (no arguments)
   run on login.

See `gtk_cleanup.py --help` for other features.


ChangeLog
0.2:

**NOTE:** A unicode-related mistake was revealed during the process of porting
from Python 2.x to Python 3.x. You may need to re-add any paths in your
blacklist which contain non-ASCII characters.

- Port from Python 2.x to 3.x
- Port from PyGTK and GTK+ 2.x to PyGObject and GTK 3.x
- Add MyPy type annotations and vim modeline to the code
- Switch from optparse to argparse for argument parsing
- Clean up the code and remove a bunch of over-engineered stuff that I can add
  back when it's actually going to get used.
- Remove --once until I can figure out how to force GTK 3 to commit the
  changes to disk without leaving the mainloop spinning indefinitely.


test_gtk_cleanup.py
#!/usr/bin/env python3

import os, shutil, tempfile, unittest, uuid

from gtk_cleanup import Blacklist, any_to_url


class TestBlacklist(unittest.TestCase):
    """Tests for the 'Blacklist' class"""
    def setUp(self):
        self.temp_dir = tempfile.mkdtemp()
        self.blacklist_path = os.path.join(self.temp_dir, 'temp.conf')
        self.bl = Blacklist(self.blacklist_path)

    def tearDown(self):
        shutil.rmtree(self.temp_dir)

    def test_hash_prefix(self):
        """Blacklist._hash_prefix: Basic function"""
        self.assertEqual(type(Blacklist._hash_prefix('test string')),
                         type('string'),
                         "_hash_method must return the standard string type")
        self.assertNotEqual(Blacklist._hash_prefix('Test String A'),
                            Blacklist._hash_prefix('Test String B'),
                            'Different strings must have different hashes')
        # TODO: Test how hashlib.sha1 behaves when given a non-ASCII path in
        # the Python 2.x version and then match that behaviour here.

    def test_hash_prefix_compat(self):
        """Blacklist._hash_prefix: Compatibility with existing blacklists"""
        self.assertEqual(Blacklist._hash_prefix('file:///bin/sh'),
            '765e4c4a771b301deb2c65918ce0e599cf7049e8',
            "_hash_method cannot change its algorithm without breaking "
            "compatibility with existing blacklists")

    def test_add_deduping(self):
        """Blacklist.add: Deduplicates additions"""
        self.assertEqual(self.bl._contents, [])
        self.bl.add('foo')
        self.assertEqual(self.bl._contents, [(3, self.bl._hash_prefix('foo'))])
        self.bl.add('foo')
        self.assertEqual(self.bl._contents, [(3, self.bl._hash_prefix('foo'))])

        self.bl.add('fooo')
        self.bl.add('foooo')
        self.bl.add('fooooo')
        self.bl.add('foooooo')
        self.assertEqual(self.bl._contents, [(3, self.bl._hash_prefix('foo'))])

    def test_index_no_match(self):
        """Blacklist.index: Raises IndexError on no match"""
        self.assertRaises(IndexError, self.bl.index, 'nonexistent')

    def test_index(self):
        """Blacklist.index: Prefix matching behaviour is correct

        (i.e. it doesn't special-case path separators because that might change
        how blacklists built under v0.1 are interpreted.)
        """
        self.assertRaises(IndexError, self.bl.index, 'file:///bin/sh')
        self.bl.add('file:///bin/sh')
        self.assertEqual(self.bl.index('file:///bin/sh'), 0)
        self.assertEqual(self.bl.index('file:///bin/shar'), 0)
        self.assertEqual(self.bl.index('file:///bin/sh/'), 0)
        self.assertEqual(self.bl.index('file:///bin/sh/foo'), 0)
        self.assertRaises(IndexError, self.bl.index, 'file:///bin/bar')

        self.assertRaises(IndexError, self.bl.index, 'file:///bin/bash/')
        self.bl.add('file:///bin/bash/')
        self.assertRaises(IndexError, self.bl.index, 'file:///bin/bash')
        self.assertRaises(IndexError, self.bl.index, 'file:///bin/bashar')
        self.assertEqual(self.bl.index('file:///bin/bash/'), 1)
        self.assertEqual(self.bl.index('file:///bin/bash/foo'), 1)

    def test_load_nonexistent(self):
        """Blacklist: Handles nonexistent blacklists gracefully"""
        self.assertFalse(Blacklist(str(uuid.uuid4())).load())

    def test_load_failures(self):
        """Blacklist: Handles miscellaneous load failures gracefully"""
        # Access Denied
        self.assertRaises(OSError,
            lambda: Blacklist("/etc/ssl/private").load())

        # Invalid UTF-8
        self.assertFalse(Blacklist("/bin/sh").load())

    def test_load_malformed(self):
        """Blacklist: Handles malformed blacklists gracefully"""
        with open(self.blacklist_path, 'w') as fobj:
            fobj.write('765e4c4a771b301deb2c65918ce0e599cf7049e8 14\n')
            fobj.flush()

            # Sanity check that load succeeds with well-formed data
            self.assertTrue(self.bl.load())

        for (data, message) in (
                    ('a 14\n', 'field 0 length is wrong'),
                    ('765e4c4a771b301deb2c65918ce0e599cf7049e8 a\n',
                        'field 1 not an int'),
                    ('765e4c4a771b301deb2c65918ce0e599cf7049e8\n',
                        'there are less than 2 fields'),
                    ('765e4c4a771b301deb2c65918ce0e599cf7049e8 14 1\n',
                        'there are more than 2 fields'),
        ):
            with open(self.blacklist_path, 'w') as fobj:
                bl = Blacklist(self.blacklist_path)
                fobj.write(data)
                fobj.flush()
                self.assertFalse(bl.load(),
                    'Should fail if {}'.format(message))

    def test_load_deduplicates(self):
        """Blacklist.load: Duplicate lines are collapsed away"""
        with open(self.blacklist_path, 'w') as fobj:
            # check that it deduplicates properly, even if data gets unsorted
            # by external means
            fobj.write('765e4c4a771b301deb2c65918ce0e599cf7049e8 14\n')
            fobj.write('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa 1\n')
            fobj.write('765e4c4a771b301deb2c65918ce0e599cf7049e8 14\n')
            fobj.flush()

        self.assertTrue(self.bl.load())
        self.assertEqual(len(self.bl._contents), 2)

    def test_load_success(self):
        self.fail("TODO: Test that load produces the expected result from "
                  "known input")

    def test_load_comments(self):
        """Blacklist.load: Allows and ignores comments and whitespace"""
        self.fail("TODO: Test how load handles comments and whitespace, "
                  "including variable whitespace between fields")
        self.fail("TODO: Test that a load() / save() cycle strips comments "
                  "since the sorting process would ruin their context.")

    def test_remove_all(self):
        """Blacklist.remove_all: Basic function"""
        # Set up and assert step 1 test data
        self.assertEqual(self.bl._contents, [])
        self.bl.add('foo')
        self.bl.add('foba')
        self.assertEqual(self.bl.index('foo'), 0)
        self.assertEqual(self.bl.index('foba'), 1)

        # Assert that removing 'foo' doesn't remove 'fob' or leave a hole
        self.bl.remove_all('foo')
        self.assertRaises(IndexError, self.bl.index, 'foo')
        self.assertEqual(self.bl.index('fobaa'), 0)

        # Assert that the test data remains sorted
        # (Sorted by the prefix lengths and then hashes, not the raw text)
        self.bl.add('foo')
        self.bl.add('fa')
        self.assertEqual(self.bl.index('fa'), 0)
        self.assertEqual(self.bl.index('fobaaa'), 2)
        self.assertEqual(self.bl.index('foo'), 1)

        # Assert that removing 'fo' doesn't remove 'foo' or 'fob'
        self.bl.remove_all('fo')
        self.assertEqual(len(self.bl._contents), 3,
            "Removing a string should not remove strings with it as a prefix")

        # But the reverse relationship does work
        self.bl.remove_all('foooooooooo')
        self.bl.remove_all('fobaaaaaaaa')
        self.assertEqual(self.bl.index('faa'), 0)
        self.assertEqual(len(self.bl._contents), 1,
            "Removing a string should remove all its prefixes")

        # Assert that removing 'fa' returns to an empty list
        self.bl.remove_all('fa')
        self.assertEqual(self.bl._contents, [])

    def test_remove_all_is_all(self):
        """Blacklist: remove_all doesn't stop at the first match"""
        self.fail("TODO: Manually build a test file with overlapping prefixes "
            "and test that remove_all removes all of them.")

    def test_save_success(self):
        self.fail("TODO: Test that save produces the expected result from "
                  "known input")


class TestMisc(unittest.TestCase):
    """Tests that don't fit in any other category"""
    def test_any_to_url(self):  # nosec
        """any_to_url: Basic Functionality"""
        old_dir = os.getcwd()

        try:
            os.chdir('/bin')
            fname = str(uuid.uuid4())  # A reliably nonexistent filename
            self.assertEqual(any_to_url('/' + fname), 'file:///' + fname,
                "Absolute paths should always be translated to URLs")
            self.assertEqual(any_to_url('sh'), 'file:///bin/sh',
                "Relative paths which exist should be translated to URLs")
            self.assertEqual(any_to_url('sh/f'), 'file:///bin/sh/f',
                "A prefix as a trailing path component shouldn't break things")
            self.assertEqual(any_to_url('file:///f'), 'file:///f',
                "URLs should pass through unchanged")
        finally:
            os.chdir(old_dir)  # pragma: nocover

if __name__ == '__main__':  # pragma: nocover
    unittest.main()


gtk_cleanup.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""A simple little tool which watches GTK's global recent files list and
removes anything that matches a hashed blacklist of URI prefixes.

--snip--

NOTE: The config file is stored in XDG_DATA_DIR for two reasons:
    1. It's probably a good idea to have it roaming.
    2. It's more likely to be mistaken for a mapping between
       content hashes and something like visit counts.

NOTE: For now, the liberal use of sort() calls is acceptable because
    Timsort takes advantage of sorted subsequences and I don't expect these
    lists to get very long.

    (And if it does, a list of 20 entries can be sorted a million times over
    in under 3 seconds on my system and reverse=True'd in under 5 seconds.)

TODO: Implement a test suite with full coverage.

TODO: Write a ``setup.py`` and an ``autostart/gtk_cleanup.desktop`` for
       potential packagers.

TODO: Performance optimizations to look into:
      - Explore optimizations to reduce the work done (hashing, etc.) for each
        entry that hasn't been added since our last scrub.
      - Find a way to avoid running the scrubber in response to the changes
        it itself made. (Ideally, a general way to stop a callback from running
        on events it itself emitted without the risk of race conditions.)

TODO: Find a future-proof way to offer an option to set the private hint on
      existing recent entries as an alternative to deleting them.
      (So they only show up in the application which added them)

TODO: Design a GUI equivalent to Chrome's view for deleting individual
      history entries.
"""

# Prevent Python 2.x PyLint from complaining if run on this
from __future__ import (absolute_import, division, print_function,
                        with_statement, unicode_literals)

__author__ = "Stephan Sokolow (deitarion/SSokolow)"
__appname__ = "GTK Recent Files Scrubber"
__version__ = "0.2"
__license__ = "GNU GPL 3.0 or later"


import hashlib, logging, os, sys
from urllib.request import pathname2url

log = logging.getLogger(__name__)  # pylint: disable=C0103


XDG_DATA_DIR = os.environ.get('XDG_DATA_HOME',
        os.path.expanduser('~/.local/share'))

import gi  # type: ignore
gi.require_version("Gtk", "3.0")
gi.require_version("Gdk", "3.0")
from gi.repository import Gdk, Gtk, GLib  # type: ignore

from typing import Dict, List, Tuple


class Blacklist(object):
    def __init__(self, filename: str):
        self.filename = filename
        self._contents = []  # type: List[Tuple[int, str]]

    @staticmethod
    def _hash_prefix(prefix: str) -> str:
        """Single location for hash-generation code."""
        return hashlib.sha1(prefix.encode('utf8')).hexdigest()

    def add(self, prefix: str):
        """Add a prefix to the hashed blacklist."""
        try:
            self.index(prefix)
            log.debug("Prefix already in the blacklist: %s", prefix)
        except IndexError:
            self._contents.append((len(prefix), self._hash_prefix(prefix)))
            self._contents.sort()  # Keep sorted for short-circuit matching

    def index(self, uri: str) -> int:
        """Find the list index of the first hashed prefix that matches 'uri'.

        Raises IndexError if nothing matches.
        """
        for pos, (prefix_len, prefix_hash) in enumerate(self._contents):
            if prefix_len > len(uri):  # Rely on the list being kept sorted
                break
            if prefix_hash == self._hash_prefix(uri[:prefix_len]):
                return pos
        raise IndexError("%s does not match any prefixes in the list" % uri)

    def load(self) -> bool:
        """Load the blacklist from the associated file

        Raises OSError or a subclass of it if the file cannot be opened.

        (Will replace any unsaved changes in memory)
        """
        if not os.path.exists(self.filename):
            log.debug("No blacklist found: %s", self.filename)
            return False

        try:
            results = set()  # Use a set to collapse duplicates

            # Intentionally die if we don't have permissions for the blacklist
            with open(self.filename, 'r') as fobj:
                # ...but UnicodeDecodeError if it's malformed, since
                # UnicodeDecodeError is subclass of ValueError
                for row in fobj:
                    row = row.strip()  # Allow leading and trailing whitespace

                    if row.startswith('#') or not row:
                        continue  # Allow comments and blank lines in the file

                    # ValueError if field count or second field type are wrong
                    digest, prefix_len = row.split(None, 2)

                    if len(digest) != 40:
                        raise ValueError("Field 0 is not a valid MD5sum")
                    results.add((int(prefix_len), digest))

            # Replace atomically and keep sorted for short-circuit matching
            self._contents = list(sorted(results))
            return True
        except ValueError as err:
            log.error("Malformed blacklist (%s):\n\t%s", self.filename, err)
            return False

    def remove_all(self, uri: str):
        """Remove all prefixes from the blacklist which match ``uri``."""
        while True:
            try:
                self._contents.pop(self.index(uri))
            except IndexError:
                break

    def save(self):
        """Save any changes to the blacklist to disk.

        Raises OSError or a subclass of it if the file cannot be opened.

        NOTE: Blacklists are stored with the fields reversed on disk and in
              reversed order so that the on-disk format will resemble some kind
              of MRU list based on the hashes of file contents.
        """
        # Intentionally die if we don't have permissions for the blacklist
        with open(self.filename, 'w') as fobj:
            for prefix_len, digest in sorted(self._contents, reverse=True):
                fobj.write('%s\t%d\n' % (digest, prefix_len))


class RecentManagerScrubber(object):
    def __init__(self, blacklist: Blacklist):
        self.blacklist = blacklist
        self.manager = Gtk.RecentManager.get_default()
        self.handler = None

    def purge(self):
        """Purge all entries from attached Recently Used lists."""
        path = self.manager.props.filename
        try:
            count = self.manager.purge_items()
            log.info("Purged %d items from %s", count, path)
        except GLib.Error as err:
            log.error("Could not purge %s: %s", path, err)

    def scrub(self, recent_manager: Gtk.RecentManager):
        """Given a Gtk.RecentManager, remove all entries in the blacklist."""
        log.debug("Cleaning MRU list...")
        found = []
        for item in recent_manager.get_items():
            uri = item.get_uri()
            try:
                self.blacklist.index(uri)
                found.append(uri)
            except IndexError:
                log.debug('Skipped %s', item.get_display_name())

        if not found:
            return

        # Remove found entries in one batch so we can show a summarized message
        # (Keeps potential log files clean and avoids leaking data into them)
        log.info("Removing %d entries", len(found))
        while found:
            try:
                recent_manager.remove_item(found.pop())
            except GLib.Error:
                log.warning("Failed to remove item. (Maybe already done)")

    def start(self):
        """Scrub the MRU list and attach a handler to watch for changes.

        Raises OSError or a subclass of it if the MRU file cannot be stat()-ed.
        """
        path = self.manager.props.filename

        # Limit the most obvious side-channel for snooping on MRU info
        if not os.stat(path).st_mode & 0o777 == 0o600:
            log.warning("Bad file permissions on MRU list. Fixing: %s", path)
            try:
                os.chmod(path, 0o600)
            except OSError:
                log.error("Failed to chmod %s", path)

        # Actually scrub and attach
        self.scrub(self.manager)
        if self.handler is None:  # Don't double-connect our signals
            self.handler = self.manager.connect('changed', self.scrub)
            log.debug("Watching MRU list: %s", path)
        else:
            log.debug("Already watched. Skipping: %s", path)


def any_to_url(path_or_url: str) -> str:
    """Helper for flexible command-line input of blacklist entries"""
    is_abs = path_or_url.startswith(os.sep)
    is_alt_abs = os.altsep and path_or_url.startswith(os.altsep)
    exists = os.path.exists(path_or_url)

    # If it's something like /path/to/porn/folder-number- meant to match only
    # some children of a given folder
    exists_prefix = os.path.exists(os.path.split(path_or_url)[0])

    if is_abs or is_alt_abs or exists or exists_prefix:
        path_or_url = 'file://' + pathname2url(os.path.abspath(path_or_url))

    # Otherwise, trust the user to know what they meant
    return path_or_url


def main():
    """The main entry point, compatible with setuptools entry points."""
    from argparse import ArgumentParser, RawDescriptionHelpFormatter
    parser = ArgumentParser(formatter_class=RawDescriptionHelpFormatter,
        description=__doc__.replace('\r\n', '\n').split('\n--snip--\n')[0])
    parser.add_argument('--version', action='version',
        version="%%(prog)s v%s" % __version__)
    parser.add_argument('-v', '--verbose', action="count",
        default=2, help="Increase the verbosity. Use twice for extra effect.")
    parser.add_argument('-q', '--quiet', action="count",
        default=0, help="Decrease the verbosity. Use twice for extra effect.")
    # Reminder: %(default)s can be used in help strings.

    resopt = parser.add_argument_group("Resident-Compatible Actions")
    resopt.add_argument('--purge', action="store_true", default=False,
        help="Purge all Recently Used entries during the initial scrub.")
    resopt.add_argument('--config', action="store", metavar="FILE",
        default=os.path.join(XDG_DATA_DIR, 'grms.conf'),
        help="Specify a non-default config file (default: %(default)s)")

    nonres = parser.add_argument_group("Non-Resident Actions")
    nonres.add_argument('-a', '--add', action="append", dest="additions",
        help="Add URI to the list of blacklisted prefixes and exit.",
        default=[], metavar="URI")
    nonres.add_argument('-r', '--remove', action="append", dest="removals",
        help="Remove prefixes from the blacklist which match URI and exit.",
        default=[], metavar="URI")
    #nonres.add_argument('--once', action="store_true", default=False,
    #    help="Don't become resident. Just scrub and exit.")

    args = parser.parse_args()

    # Set up clean logging to stderr
    log_levels = [logging.CRITICAL, logging.ERROR, logging.WARNING,
              logging.INFO, logging.DEBUG]
    args.verbose = min(args.verbose - args.quiet, len(log_levels) - 1)
    args.verbose = max(args.verbose, 0)
    logging.basicConfig(level=log_levels[args.verbose],
                format='%(levelname)s: %(message)s')

    # Prepare the blacklist
    blist = Blacklist(args.config)
    blist.load()

    # TODO: Watch the config file for changes so the watcher doesn't need to be
    #       restarted manually to pick them up.
    if args.additions or args.removals:
        for arg in args.additions:
            blist.add(any_to_url(arg))
        for arg in args.removals:
            blist.remove_all(any_to_url(arg))
        blist.save()
        sys.exit(0)

    # Scrub and start watching the recent list for updates
    scrubber = RecentManagerScrubber(blist)
    scrubber.start()

    if args.purge:
        scrubber.purge()

    #if args.once:
    #    Gtk.main_iteration_do(True)
    #else:
    Gtk.main()

if __name__ == '__main__':  # pragma: nocover
    main()

# vim: set sw=4 sts=4 expandtab :
